File: code/1657.determine-if-two-strings-are-close.php
<?php
/*
 * @lc app=leetcode id=1657 lang=php
 *
 * [1657] Determine if Two Strings Are Close
 */

// @lc code=start
class Solution {

    /**
     * @param String $word1
     * @param String $word2
     * @return Boolean
     */
    function closeStrings($word1, $word2) {
        $countMap1 = array_count_values(str_split($word1));
        $countMap2 = array_count_values(str_split($word2));
        if (!empty(array_diff($countMap1, $countMap2))) return false;
        if (!empty(array_diff($countMap2, $countMap1))) return false;

        $keyMap1 = array_keys($countMap1);
        $keyMap2 = array_keys($countMap2);
        if (!empty(array_diff($keyMap1, $keyMap2))) return false;
        if (!empty(array_diff($keyMap2, $keyMap1))) return false;

        $countMap1 = array_values($countMap1);
        $countMap2 = array_values($countMap2);
        sort($countMap1);
        sort($countMap2);
        return $countMap1 == $countMap2;
    }
}
// @lc code=end

File: code/1657.determine-if-two-strings-are-close.go
package code

import (
	"slices"
)

/*
 * @lc app=leetcode id=1657 lang=golang
 *
 * [1657] Determine if Two Strings Are Close
 */

// @lc code=start
func closeStrings(word1 string, word2 string) bool {
	if len(word1) != len(word2) {
		return false
	}

	hashMap1 := make(map[rune]int)
	for _, s := range word1 {
		hashMap1[s]++
	}

	hashMap2 := make(map[rune]int)
	for _, s := range word2 {
		hashMap2[s]++
	}

	return slices.Equal(convertSliceKeys(hashMap1), convertSliceKeys(hashMap2)) && slices.Equal(convertSliceValues(hashMap1), convertSliceValues(hashMap2))
}

func convertSliceKeys(m map[rune]int) []rune {
	result := make([]rune, 0, len(m))
	for k := range m {
		result = append(result, k)
	}
	slices.Sort(result)
	return result
}

func convertSliceValues(m map[rune]int) []int {
	result := make([]int, 0, len(m))
	for _, v := range m {
		result = append(result, v)
	}
	slices.Sort(result)
	return result
}

// @lc code=end
